File: clojure/src/aoc/day2.clj
(ns aoc.day2
  (:require [clojure.string :as string]
            [aoc.util :as util]))

(def input "./../inputs/day2.txt")

(def line-regex #"(?<min>\d+)-(?<max>\d+) (?<char>.): (?<password>.+)")

(defn- parse-line [line]
  (let [matcher (re-matcher line-regex line)]
    (.matches matcher)
    {:password (.group matcher "password")
     :char (->> "char" (.group matcher) first)
     :min (->> "min" (.group matcher) util/str->int)
     :max (->> "max" (.group matcher) util/str->int)}))

(defn- read-input []
  (->>
   input
   (slurp)
   (string/split-lines)
   (map parse-line)))

(defn- part-1-is-valid [{:keys [password char min max]}]
  (->>
   password
   (filter #(= char %))
   (count)
   (#(<= min % max))))

(defn run-part-1 []
  (->>
   (read-input)
   (filter #(part-1-is-valid %))
   (count)))

(defn- part-2-is-valid [{:keys [password char min max]}]
  (->>
   [(get password (dec min)) (get password (dec max))]
   (filter #(= char %))
   (count)
   (#(= 1 %))))

(defn run-part-2 []
  (->>
   (read-input)
   (filter #(part-2-is-valid %))
   (count)))

File: haskell/src/Day2.hs
module Day2
  ( runPart1,
    runPart2,
  )
where

import qualified Data.Either.Unwrap
import qualified Text.Parsec as Parsec
import Util

input :: String
input = "../inputs/day2.txt"

type Password = String

data Rule = Rule Int Int deriving (Show)

data PasswordAndRule = PasswordAndRule Password Rule Char deriving (Show)

parser = do
  min <- Parsec.many1 Parsec.digit
  _ <- Parsec.string "-"
  max <- Parsec.many1 Parsec.digit
  _ <- Parsec.string " "
  char <- Parsec.anyChar
  _ <- Parsec.string ": "
  password <- Parsec.many1 Parsec.anyChar
  return $ PasswordAndRule password (Rule (read min) (read max)) char

parsePasswordAndRule :: String -> PasswordAndRule
parsePasswordAndRule line =
  line
    |> Parsec.parse parser ""
    |> Data.Either.Unwrap.fromRight

readInput :: IO [PasswordAndRule]
readInput = do
  fileContents <- readFile input
  fileContents
    |> lines
    |> map parsePasswordAndRule
    |> pure

solvePart1 :: [PasswordAndRule] -> String
solvePart1 rules = rules |> filter isValid |> length |> show
  where
    isValid (PasswordAndRule password (Rule min max) char) =
      password
        |> filter (== char)
        |> length
        |> (\count -> (count >= min) && (count <= max))

solvePart2 :: [PasswordAndRule] -> String
solvePart2 rules = rules |> filter isValid |> length |> show
  where
    isValid (PasswordAndRule password (Rule min max) char) =
      [password !! (min - 1), password !! (max - 1)]
        |> filter (== char)
        |> length
        |> (== 1)

runPart1 :: IO ()
runPart1 = do
  rules <- readInput
  rules |> solvePart1 |> putStrLn

runPart2 :: IO ()
runPart2 = do
  rules <- readInput
  rules |> solvePart2 |> putStrLn